# Day 4 — Trees, Forests, SVMs & k-NN

## Overview

In this day, we will explore and compare different machine learning classification algorithms, including Decision Trees, Random Forests, Support Vector Machines (SVM), and K-Nearest Neighbors (K-NN).

The goal is to train these models on the same dataset, evaluate their performance using the F1-score metric, compare their results, and identify the best-performing model.

Additionally, we will analyze the feature importance generated by the Random Forest model to understand which features have the greatest impact on predictions.

In [34]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier



## Step 1: Load the Dataset

The Titanic dataset is loaded using pandas.

This dataset contains information about passengers, including demographic details, ticket information, and survival status.

The goal is to use passenger features to predict whether a passenger survived or not.

In [3]:
df = pd.read_csv("Titanic-Dataset.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Step 2: Data Cleaning and Feature Selection

Before building the model, the dataset is cleaned and prepared.

### Initial Missing Values Check

First, the dataset is checked for missing values to identify columns that require preprocessing.

This step helps determine which features contain missing data and need to be handled before training the model.

### Selecting Relevant Features

Several columns in the original dataset do not provide useful information for prediction, such as passenger identifiers, names, ticket numbers, and cabin details.

Only relevant features are selected:
- Passenger class (`Pclass`)
- Gender (`Sex`)
- Age (`Age`)
- Family relationships (`SibSp`, `Parch`)
- Fare (`Fare`)
- Embarkation location (`Embarked`)

The unnecessary columns are removed to keep only meaningful features.

### Handling Missing Values

Missing values are handled using:
- Median value for missing `Age` values.
- Most frequent category for missing `Embarked` values.

### Encoding Categorical Features

Categorical variables are converted into numerical values because Logistic Regression requires numerical input.

The encoding process is:
- Male = 0, Female = 1
- S = 0, C = 1, Q = 2

In [4]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [5]:
# Select the features that will be used in the model
df = df[["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked", "Survived"]]

# Fill missing values
df["Age"] = df["Age"].fillna(df["Age"].median())
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode().iloc[0])

# Encode categorical variables using mapping
df["Sex"] = df["Sex"].map({"male": 0, "female": 1})
df["Embarked"] = df["Embarked"].map({"S": 0, "C": 1, "Q": 2})

### Verifying Missing Values After Preprocessing

After handling the missing values, the dataset is checked again using `isnull().sum()` to confirm that all missing values have been successfully handled.


In [6]:
df.isnull().sum()

Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
Survived    0
dtype: int64

## Step 4: Separate Features and Target

The dataset is divided into:

- **Features (X):** The input variables used by the model to make predictions.
- **Target (y):** The output variable that represents passenger survival.

The `Survived` column is separated as the target, while all remaining columns are used as features.

In [7]:
X = df.drop("Survived", axis=1)
y = df["Survived"]

## Step 5: Split the Dataset

The dataset is divided into training and testing sets using an 80/20 split.

The training data is used to train the Logistic Regression model, while the testing data is used to evaluate the model's ability to generalize to unseen data.

A fixed `random_state` is used to ensure reproducible results.

In [8]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(712, 7)
(179, 7)


## Step 6: Training and Evaluating Classification Models

In this step, we will train four different classification models using the same training and testing data split to ensure a fair comparison.

The models that will be evaluated are:
- Decision Tree
- Random Forest
- Support Vector Machine (SVM)
- K-Nearest Neighbors (K-NN)

Each model will be trained using the training data, then used to make predictions on the test data. After that, the **F1-score** will be calculated for each model to evaluate and compare their performance.

The F1-score was selected as the evaluation metric because it provides a balance between precision and recall, making it suitable for classification problems.



### 6.1 Decision Tree Classifier

The Decision Tree model was trained using the training dataset and used to predict the target values on the test dataset.

The F1-score was calculated to evaluate the model's performance and measure the balance between precision and recall.

In [ ]:
tree_model = DecisionTreeClassifier(max_depth=5 , random_state=42)
tree_model.fit(X_train,y_train)

tree_pred = tree_model.predict(X_test)

In [17]:
tree_f1= f1_score(y_test,tree_pred )
print(tree_f1)

0.7272727272727273


### 6.2 Random Forest Classifier

The Random Forest model was trained using multiple decision trees to improve prediction performance and reduce overfitting.

The model predictions were generated on the test dataset, and the F1-score was calculated to evaluate its classification performance.

In [16]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

rf_pred = rf_model.predict(X_test)

In [18]:
rf_f1= f1_score(y_test,rf_pred )
print(rf_f1)

0.7832167832167832


### 6.3 Support Vector Machine (SVM)

The Support Vector Machine model was trained to find the optimal decision boundary between different classes.

After training, predictions were generated for the test dataset, and the F1-score was calculated to evaluate the model performance.

In [31]:
svm_model = SVC(random_state=42)
svm_model.fit(X_train, y_train)

svm_pred = svm_model.predict(X_test)

In [ ]:
svm_f1= f1_score(y_test,svm_pred )
print(svm_f1)

0.3673469387755102


### 6.4 K-Nearest Neighbors (KNN)

The K-Nearest Neighbors model was trained based on the similarity between data points.

The model was used to predict the test data, and the F1-score was calculated to measure its classification performance.

In [21]:
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)

knn_pred = knn_model.predict(X_test)

In [22]:
knn_f1= f1_score(y_test,knn_pred )
print(knn_f1)

0.6046511627906976


## Step 7: Model Performance Comparison

In this step, the F1-score results of all trained models will be collected and organized into a single comparison table.

This table allows us to compare the performance of the Decision Tree, Random Forest, SVM, and K-NN models using the same evaluation metric.

By comparing the F1-scores, we can identify which model achieved the best performance on this dataset and select the most suitable model for prediction.

In [28]:
results = pd.DataFrame({
    "Model": ["Decision Tree", "Random Forest", "SVM", "KNN"],
    "F1-score": [tree_f1, rf_f1, svm_f1, knn_f1]
})

results

,Model,F1-score
0,Decision Tree,0.727273
1,Random Forest,0.783217
2,SVM,0.367347
3,KNN,0.604651


## Interpretation of Model Comparison Results

Based on the F1-score results, the **Random Forest model achieved the highest performance with an F1-score of 0.783217**, making it the best-performing model among the evaluated classifiers.

The Decision Tree model achieved a good F1-score of **0.727273**, but its performance was lower than Random Forest because it relies on a single decision tree, which can be more sensitive to overfitting.

The K-NN model achieved a moderate F1-score of **0.604651**. Its performance may be affected by the distance calculations between samples and the distribution of the features.

The SVM model achieved the lowest F1-score of **0.367347**, which indicates that it was not able to separate the classes effectively with the current configuration. This could be improved by applying feature scaling or tuning the model parameters.

## Step 8: Random Forest Feature Importance Analysis

In this step, we analyze the importance of each feature in the Random Forest model to understand which variables have the greatest impact on the model's predictions.

The `feature_importances_` attribute calculates the contribution of each feature to the decision-making process of the Random Forest model.

The features are sorted in descending order, where the features with higher importance scores have a stronger influence on the model predictions, while features with lower scores contribute less.

This analysis helps us understand the relationship between the input features and the target variable, and it can also be useful for future feature selection and model improvement.

In [ ]:

# Create a DataFrame containing each feature and its importance score from the Random Forest model

feature_importance = pd.DataFrame({
    "Feature": X.columns,
    "Importance": rf_model.feature_importances_
})

# This places the most influential features at the top
feature_importance = feature_importance.sort_values(
    by="Importance",
    ascending=False
)

print(feature_importance)

    Feature  Importance
5      Fare    0.269744
1       Sex    0.268418
2       Age    0.256426
0    Pclass    0.081891
3     SibSp    0.052077
4     Parch    0.037390
6  Embarked    0.034054


## Interpretation of Random Forest Feature Importance Results

The feature importance analysis shows the contribution of each feature to the Random Forest model's predictions.

According to the results, **Fare** was the most important feature with an importance score of **0.269744**, followed closely by **Sex** with **0.268418** and **Age** with **0.256426**. This indicates that these three features had the strongest influence on the model's decision-making process.

The **Pclass** feature also contributed to the predictions with an importance score of **0.081891**, while **SibSp**, **Parch**, and **Embarked** had lower importance values, meaning they had a smaller impact on the model's predictions.

These results suggest that passenger information such as ticket fare, gender, and age were the main factors used by the Random Forest model to predict the target variable. The analysis also helps explain how the model makes decisions and identifies the most influential features in the dataset.

## Model Performance Comparison

Based on the evaluation results shown in the previous table, the four models were compared using the F1-score metric on the same train/test split.

The results indicate that the **Random Forest model achieved the highest F1-score (0.783217)** compared with Decision Tree (0.727273), K-NN (0.604651), and SVM (0.367347). Therefore, Random Forest was the best-performing model for this dataset.

Random Forest achieved better performance because it combines multiple decision trees and uses their aggregated predictions, which helps reduce overfitting and improves the model's ability to generalize to unseen data.

Although the Decision Tree achieved a good score, it relies on a single tree, which can make it more sensitive to overfitting. K-NN showed moderate performance but can be affected by the distance between samples and feature distribution. SVM achieved the lowest score, which may indicate that it requires further parameter tuning or feature scaling to achieve better results.

Based on these results, **Random Forest was selected as the final model because it provided the best performance and the most reliable predictions among the tested models.**